documentation/specs/src/economics/fee-system.md

```diff
@@ -1,7 +1,143 @@
+
+1. Users are responsible for correctly timing/funding their transactions with the help of the wallet
+2. We force in protocol that a block should list the wrappers after the decrypted transactions
+
+If we follow the second option the block proposers will no more be able to optimize the block (this would require running the inner transactions to calculate the possibly new unshielded balance) and, inevitably, some wrapper transactions for which fees cannot be paid will end up in the block. These will be deemed invalid during validation so that the corresponding inner transaction will not be executed, preserving the correctness of the state machine, but it represents a slight underoptimization of the block and a potential vector for DOS attacks since the invalid wrapper has allocated space and gas in the block without being charged due to the lack of funds. Because of this, we stick to the first option by not imposing any specific order in procotol.
+
+Fees are collected via protocol, in the `finalize_block` function, for `WrapperTx`s which have been processed with success: this is to prevent a malicious block proposer from including transactions that are known in advance to be invalid just to collect more fees. Given the two-block execution model of Namada (wrapper and inner tx) and the need to collect fees for the allocated resources, nothing can be done in case the inner transaction fails: by that point, fees have already been collected and no refunds will be issued, meaning that the inner tx signer is responsible for submitting a semantically valid transaction for the state of the application (importance on the lifetime parameter of the tx here). If enough funds are available, these are deducted from the unshielded storage balances of the fee payers and directed to the balance of the block proposer: payers (wrapper tx signers) are responsible for keeping enough unshielded funds for their transactions. If instead, the balance is not enough to cover fees, then the corresponding inner transaction gets discarded and no funds are moved from the payer address to the block proposer. This is due to the following reasons:
+
+- It would penalize the tx submitter who might not be responsible for the lack of funds at that moment
+- Moving insufficient funds would incentivize the block proposer to include transactions for which fees cannot be paid. Since the block proposer knows the balances of the involved addresses at block creation time and given the strategy of placing wrapper txs first, this would constitute malicious behavior by the proposer
+
+By discarding the transaction without paying fees instead we avoid these pitfalls. This logic implies that the strategy of placing wrapper transactions before any decrypted tx in the block will be reinforced. It might look like a contradiction to what was said before, a wrapper transaction included in a block will not pay fees for the resources it acquired, but this is not true:
```

Comment:
including a transaction which cannot pay fees should be considered an invalid block proposal